File: app/src/main/java/me/wcy/music/service/AudioFocusManager.java
package me.wcy.music.service;

import android.content.Context;
import android.media.AudioManager;

import static android.content.Context.AUDIO_SERVICE;

/**
 * Created by hzwangchenyan on 2017/8/8.
 */
public class AudioFocusManager implements AudioManager.OnAudioFocusChangeListener {
    private AudioManager audioManager;
    private boolean isPausedByFocusLossTransient;

    public AudioFocusManager(Context context) {
        audioManager = (AudioManager) context.getSystemService(AUDIO_SERVICE);
    }

    public boolean requestAudioFocus() {
        return audioManager.requestAudioFocus(this, AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN)
                == AudioManager.AUDIOFOCUS_REQUEST_GRANTED;
    }

    public void abandonAudioFocus() {
        audioManager.abandonAudioFocus(this);
    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        switch (focusChange) {
            // 重新获得焦点
            case AudioManager.AUDIOFOCUS_GAIN:
                if (isPausedByFocusLossTransient) {
                    // 通话结束，恢复播放
                    AudioPlayer.get().startPlayer();
                }

                // 恢复音量
                AudioPlayer.get().getMediaPlayer().setVolume(1f, 1f);

                isPausedByFocusLossTransient = false;
                break;
            // 永久丢失焦点，如被其他播放器抢占
            case AudioManager.AUDIOFOCUS_LOSS:
                AudioPlayer.get().pausePlayer();
                break;
            // 短暂丢失焦点，如来电
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                AudioPlayer.get().pausePlayer(false);
                isPausedByFocusLossTransient = true;
                break;
            // 瞬间丢失焦点，如通知
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                // 音量减小为一半
                AudioPlayer.get().getMediaPlayer().setVolume(0.5f, 0.5f);
                break;
        }
    }
}

File: app/src/main/java/me/wcy/music/service/AudioFocusManager.kt
package me.wcy.music.service

import android.content.Context
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener

/**
 * Created by hzwangchenyan on 2017/8/8.
 */
class AudioFocusManager(
    private val context: Context,
    private val audioPlayer: AudioPlayer
) {
    private val audioManager by lazy {
        context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }
    private var isPausedByFocusLossTransient = false

    fun requestAudioFocus(): Boolean {
        return (audioManager.requestAudioFocus(
            focusChangeListener,
            AudioManager.STREAM_MUSIC,
            AudioManager.AUDIOFOCUS_GAIN
        ) == AudioManager.AUDIOFOCUS_REQUEST_GRANTED)
    }

    fun abandonAudioFocus() {
        audioManager.abandonAudioFocus(focusChangeListener)
    }

    private val focusChangeListener =
        OnAudioFocusChangeListener { focusChange ->
            when (focusChange) {
                AudioManager.AUDIOFOCUS_GAIN -> {
                    if (isPausedByFocusLossTransient) {
                        // 通话结束，恢复播放
                        audioPlayer.startPlayer()
                    }

                    // 恢复音量
                    audioPlayer.setVolume(1f, 1f)
                    isPausedByFocusLossTransient = false
                }

                AudioManager.AUDIOFOCUS_LOSS -> {
                    audioPlayer.pausePlayer()
                }

                AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                    audioPlayer.pausePlayer(false)
                    isPausedByFocusLossTransient = true
                }

                AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                    // 音量减小为一半
                    audioPlayer.setVolume(0.5f, 0.5f)
                }
            }
        }
}